rfcs/20221010-add-ocl-ooo-queue/README.md

```diff
@@ -0,0 +1,125 @@
+///     #DNNL_RUNTIME_DIM_VAL), the memory object must have the same memory
+///     descriptor as that returned by
+///     #dnnl_primitive_desc_query_md(#dnnl_query_exec_arg_md, index).
+/// @param deps A pointer to std::vector<cl_event> that contains
+///     dependencies.
+/// @param return_event Output event. The library doesn't own the event.
+///     It is the user's responsibility to manage lifetime of the event.
+/// @returns #dnnl_success on success and a status describing the error
+///     otherwise.
+dnnl_status_t dnnl_ocl_interop_primitive_execute(
+        const_dnnl_primitive_t primitive, dnnl_stream_t stream, int nargs,
+        const dnnl_exec_arg_t *args, const cl_event *deps, int ndeps,
+        cl_event *return_event);
+```
+
+#### C++ API
+```cpp
+// dnnl_ocl.hpp
+
+/// Executes computations specified by the primitive in a specified stream and
+/// returns an OpenCL event.
+///
+/// Arguments are passed via an arguments map containing
+/// <index, memory object> pairs. The index must be one of the `DNNL_ARG_*`
+/// values such as `DNNL_ARG_SRC`, and the memory must have a memory descriptor
+/// matching the one returned by
+/// #dnnl::primitive_desc::query_md(#query::exec_arg_md, index) unless using
+/// dynamic shapes (see #DNNL_RUNTIME_DIM_VAL).
+///
+/// @param aprimitive Primitive to execute.
+/// @param astream Stream object. The stream must belong to the same engine
+///     as the primitive.
+/// @param args Arguments map.
+/// @param deps Optional vector with `cl::sycl::event` dependencies.
+///
+/// @returns Output event. The library doesn't own the event. It is the user's
+///     responsibility to manage lifetime of the event.
+inline cl_event execute(const dnnl::primitive &aprimitive,
+        const stream &astream, const std::unordered_map<int, memory> &args,
+        const std::vector<cl_event> &deps = {});
+```
+
+### Generic API
+Technically, the generic API for primitive execution is available for any stream. However, using
+such API for an out-of-order stream doesn't make much sense because a user cannot specify
+input dependencies and get an output event to embed oneDNN primitives in a larger
+dependencies graph.
+
+There is a need to specify behavior of the generic execute API for an out-of-order stream.
+There are a couple of options.
+
+#### Option 1: Prohibit Using Generic Execute API
+The idea is to check whether stream is in-order or out-of-order in the generic execute API and
+return an `invalid_arguments` status in the latter case.
+
+Pros:
+* It is clearly defined that the generic API is not suitable for out-of-order execution model
+* Less error prone implementation because there is no need to handle generic and ocl interop APIs
+differently inside the library
+
+Cons:
+* Discrepancy in behavior of SYCL and OCL runtimes. The former one allows to use the generic
+execute API
+
+#### Option 2: Allow to Use Only Generic or Only Interop Execute API
```

Comment:
I am not sure I fully understand option 2. If we let the stream carry dependencies between execute call, it will just emulate an in-order behavior of execute calls. The only kernel overlap that would then happen would be those submitted internally within an execute call.

Is that the expectation? If so, I believe we don't need to carry any dependency in stream, we could simply enqueue a barrier at the beginning and end of the generic execute call for out of order queues.
This would also allow to mix generic and event-based execute functions. What do you think?

Reply:
>  If we let the stream carry dependencies between execute call, it will just emulate an in-order behavior of execute calls.

This is the intention. Users can use either interop API or generic API but won't be allowed to mix them. But given that OpenVINO actually may want to use generic API + barriers and interop API with events the option is no longer valid.
I think we will have to allow mixing the generic and interop APIs but without doing any synchronization to allow users to do it as they want. If we have to support both APIs then I agree with Eugene's point that we need to be as close to the OpenCL programming model as possible.

> if the queue is created outside of the plugin, and passed to the plugin.

Here I think it's the plugin's responsibility to check whether it supports the given queue or not.


> The only kernel overlap that would then happen would be those submitted internally within an execute call.

Just want to clarify that execution of nested kernels will always be in-order.